tangelo/toolboxes/operators/trim_trivial_qubits.py

```diff
@@ -0,0 +1,245 @@
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import numpy as np
+
+from tangelo.toolboxes.operators import QubitOperator, count_qubits
+from tangelo.linq import Circuit
+from tangelo.linq.helpers.circuits import pauli_string_to_of, pauli_of_to_string
+
+
+def trim_trivial_operator(qu_op, trim_states, n_qubits=None, reindex=True):
+    """
+    Calculate expectation values of a QubitOperator acting on qubits in a
+    trivial |0> or |1> state. Return a trimmed QubitOperator with updated coefficients
+
+    Args:
+        qu_op (QubitOperator): Operator to trim
+        trim_states (dict): Dictionary mapping qubit indices to states to trim, e.g. {1: 0, 3: 1}
+        n_qubits (int): Optional, number of qubits in full system
+        reindex (bool): Optional, if True, remaining qubits will be reindexed
+    Returns:
+        QubitOperator : trimmed QubitOperator with updated coefficients
+    """
+
+    qu_op_trim = QubitOperator()
+    n_qubits = count_qubits(qu_op) if n_qubits is None else n_qubits
+
+    # Calculate expectation values of trivial qubits, update coefficients
+    for op, coeff in qu_op.terms.items():
+        term = pauli_of_to_string(op, n_qubits)
+        c = np.ones(len(trim_states))
+        new_term = term
+        for i, qubit in enumerate(trim_states.keys()):
+            if term[qubit] in {'X', 'Y'}:
+                c[i] = 0
+            elif (term[qubit], trim_states[qubit]) == ('Z', 1):
+                c[i] = -1
+
+            new_term = new_term[:qubit - i] + new_term[qubit - i + 1:] if reindex else new_term[:qubit] + 'I' + new_term[qubit + 1:]
+
+        qu_op_trim += np.prod(c) * coeff * QubitOperator(pauli_string_to_of(new_term))
+    return qu_op_trim
+
+
+def is_bitflip_gate(gate):
+    """
+    Check if a gate is a bitflip gate.
+
+    A gate is a bitflip gate if it satisfies one of the following conditions:
+    1. The gate name is either X, Y.
+    2. The gate name is RX or RY, and has a parameter that is an odd multiple of pi.
+
+    Args:
+        gate (Gate): The gate to check.
+
+    Returns:
+        bool: True if the gate is a single qubit bitflip gate, False otherwise.
+    """
+    if gate is None:
+        return False
+
+    if gate.name in {"X", "Y"}:
+        return True
+    elif gate.name in {"RX", "RY"}:
+        try:
+            parameter_float = float(gate.parameter)
+        except (TypeError, ValueError):
+            return False
+
+        return abs(parameter_float - round(parameter_float / np.pi) * np.pi) < 1e-5 and round(parameter_float / np.pi) % 2 == 1
```

Comment:
The whole thing is checking if it is (or is close to) an odd multiple of pi. I did it this way because I struggled to get the modulo operator to work with a parameter close to pi like 3.14159 and also a multiple of pi. But the following seems to work!

`return abs(parameter_float % (3.14159 * 2) - 3.14159) <= 1e-5`

Reply:
Oh yeah, this this would have been my suggestion then (except I would have used np.pi ?). Regarding the "seems" to work: I'd suggest trying different value and making sure that level of precision (1e-5) makes sense to you. If you believe this is working as intended then it's all good and you can resolve this.

I know that hardware has limitations (I recall IonQ can only apply angles up to 1e-3 precision), if you think adding an optional parameter for letting the user set the tolerance is a good idea, I suggest you do it now.

Regardless, please feel free to resolve this when you are satisfied with the behaviour.